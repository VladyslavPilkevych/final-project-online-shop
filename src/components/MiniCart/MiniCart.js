/* eslint-disable no-underscore-dangle */
/* eslint-disable max-len */
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import { toggleCart, deleteFromCart } from '../../store/actionCreators/cartAC';
import { ReactComponent as CloseCartIcon } from '../../assets/icons/closeCartIcon.svg';
import numberWithSpaces from '../../utils/numberWithSpaces';

import Image from '../Image/Image';

import styles from './MiniCart.module.scss';

function MiniCart() {
  const isOpenCart = useSelector((state) => state.cart.isOpenCart);
  const dataCart = useSelector((state) => state.cart.dataCart);
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const cartItem = dataCart || [];

  const totalPrice = numberWithSpaces(cartItem?.map((item) => item?.product?.currentPrice * item?.cartQuantity).reduce((acc, value) => acc + value, 0));

  if (!isOpenCart) return null;

  const closeCart = () => {
    dispatch(toggleCart(false));
  };

  const onDeleteFromCart = (productId) => {
    if (cartItem?.map((item) => item?.product._id === productId)) {
      dispatch(deleteFromCart(productId));
    }
  };

  return (
    <div className={styles.miniCartWrapper} role="button" tabIndex="0" onClick={closeCart}>
      <div className={styles.miniCart} role="button" tabIndex="0" onClick={(e) => e.stopPropagation()}>
        <div className={styles.miniCartHeader}>
          <h3 className={styles.miniCartTitle}>My Cart</h3>
          <p className={styles.miniCartSubTitle}>
            {cartItem.length}
            -items
          </p>
          <button
            type="button"
            className={styles.editBtn}
            onClick={() => {
              navigate('/cart');
              dispatch(toggleCart(!isOpenCart));
            }}
          >
            View or Edit Your Cart
          </button>
        </div>
        <div className={styles.miniCartContent} role="button" tabIndex="0" onClick={(e) => e.stopPropagation()}>
          <ul>
            {cartItem?.map((item) => (
              <li key={item?._id}>
                <div className={styles.miniCartContentwrapper}>
                  <div>
                    <p className={styles.miniCartQuantity}>
                      {item?.cartQuantity}
                      <span>x</span>
                    </p>
                  </div>
                  {cartItem && (
                    <div
                      onClick={() => {
                        navigate(`/products/${item?.product?.itemNo}`);
                        dispatch(toggleCart(!isOpenCart));
                      }}
                      role="button"
                      tabIndex="0"
                    >
                      <Image className={styles.imageInCart} src={item?.product?.imageUrls[0]} alt={item?.product?.name + item?.product?.model} />
                    </div>
                  )}
                  <div>
                    <p className={styles.miniCartDescription}>{item?.product?.description}</p>
                  </div>
                  <div>
                    <CloseCartIcon onClick={() => onDeleteFromCart(item?.product?._id)} />
                  </div>
                </div>
              </li>
            ))}
          </ul>
        </div>
        <div className={styles.miniCartFotter}>
          <p className={styles.miniCartTotalPrice}>
            Subtotal:
            <span className={styles.miniCartTotalValues}>
              $
              {totalPrice}
            </span>
          </p>
        </div>
      </div>
    </div>
  );
}

export default MiniCart;
