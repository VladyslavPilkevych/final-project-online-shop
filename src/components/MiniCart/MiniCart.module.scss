@import '../../assets/styles/global.module.scss';
.miniCart {
  display: flex;
  flex-direction: column;
  align-items: center;
  position: absolute;
  z-index: 25;
  right: 10px;
  top: 67px;
  width: 310px;
  background-color: $white;
  border: 1px solid $grey-star;
  box-shadow: $shadow-1;
  padding: 25px 15px 25px 25px;
  box-sizing: border-box;
}
.miniCart::after {
  content: ' ';
  position: absolute;
}
.miniCartHeader {
  display: flex;
  flex-direction: column;
  align-items: center;
}
.miniCartTitle {
  font-weight: 600;
  font-size: 18px;
  line-height: 140%;
  color: $black;
  //   margin-top: 23px;
  margin-bottom: 7px;
}
.miniCartSubTitle {
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
  color: $dark-grey-text;
  margin-bottom: 18px;
}
.editBtn {
  font-weight: 600;
  font-size: 14px;
  line-height: 21px;
  border: 2px solid #0156ff;
  box-sizing: border-box;
  border-radius: 50px;
  color: $blue;
  padding: 8px 49px;
  background-color: $white;
  margin-bottom: 23px;
  cursor: pointer;
}
.miniCartContentwrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  border-top: 1px solid $grey-star;
  // border-bottom: 1px solid $grey-star;
  box-sizing: border-box;
  padding-top: 17px;
  padding-bottom: 17px;
  cursor: pointer;
}
.miniCartFotter {
  width: 268px;
  border-top: 1px solid $grey-star;
  margin: 0 auto;
}
.miniCartTotalPrice {
  margin-top: 17px;
  text-align: center;
  font-weight: 600;
  font-size: 14px;
  line-height: 140%;
  color: $dark-grey-text;
}
.miniCartTotalValues {
  font-weight: 600;
  font-size: 18px;
  line-height: 140%;
  color: $black;
}
