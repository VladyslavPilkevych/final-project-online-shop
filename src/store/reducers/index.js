import { combineReducers } from 'redux';
import preloaderReducer from './preloaderReducer';
import productsReducer from './productsReducer';
import menuReducer from './menuReducer';
import userReducer from './userReducer';
import searchReducer from './searchReducer';
import cartReducer from './cartReducer';
import miniMenuReducer from './miniMenuReducer';
import filtersCategoriesReducer from './filtersCategoriesReducer';

const reducer = combineReducers({
  loader: preloaderReducer,
  products: productsReducer,
  menu: menuReducer,
  miniMenu: miniMenuReducer,
  filtersCategories: filtersCategoriesReducer,
  user: userReducer,
  search: searchReducer,
  cart: cartReducer,
});

export default reducer;
