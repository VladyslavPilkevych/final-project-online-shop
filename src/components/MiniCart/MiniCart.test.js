import { render } from '@testing-library/react';
import MiniCart from './MiniCart';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';

function Component() {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <MiniCart />
      </Provider>
    </BrowserRouter>
  );
}

describe('Snapshot MiniCart should render', () => {
  test('Snapshot MiniCart is rendered', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
});