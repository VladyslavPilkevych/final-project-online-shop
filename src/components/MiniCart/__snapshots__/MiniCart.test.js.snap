// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot MiniCart should render Snapshot MiniCart is rendered 1`] = `<DocumentFragment />`;
